### ISSUE TITLE
PAM50 atlas warped to the subject's space gives corrupted values

### ISSUE THREAD

**Issue by jcohenadad** [spinalcordtoolbox/spinalcordtoolbox#3843]
### Description

This issue was originally reported here: https://github.com/sct-pipeline/gm-challenge/issues/77

The white matter atlas is supposedly encoded between 0 and 1. After estimating warping fields between the subject's space and the PAM50 space, those warping fields are applied to the PAM50 objects (including the WM atlas), and the values should _still_ range between 0 and 1, which does not seem to be the case in this reported bug (one voxel had value `-1`). 

### Suggestions

1. Make sure that the WM atlas has ALL values between 0 and 1. 
2. Reproduce the bug and... find a solution 😉  (likely something got corrupted with the warping field)

**Comment by valosekj** [comment]
Just for clarification (I am the author of the original issue https://github.com/sct-pipeline/gm-challenge/issues/77); I did not work with PAM50 labels. Instead, I segmented the SC and GM using `sct_deepseg_sc` and `sct_deepseg_gm` functions.

## Steps to reproduce the bug:

Download data:
[sct_issue_3843.zip](https://github.com/spinalcordtoolbox/spinalcordtoolbox/files/9109177/sct_issue_3843.zip)

Run:
```
# Segment spinal cord
sct_deepseg_sc -i sub-001_T2star.nii.gz -c t2s

# Segment gray matter
sct_deepseg_gm -i sub-001_T2star.nii.gz

# Generate white matter segmentation
sct_maths -i sub-001_T2star_seg.nii.gz -sub sub-001_T2star_gmseg.nii.gz -o sub-001_T2star_wmseg.nii.gz

# Erode white matter mask to minimize partial volume effect
sct_maths -i sub-001_T2star_wmseg.nii.gz -erode 1 -o sub-001_T2star_wmseg_erode.nii.gz

# Compute SNR, CNR, and contrast
# (the compute_snr script is from the gm-challenge package) 
compute_cnr --data1 sub-001_T2star.nii.gz --data2 .nii.gz --mask-noise sub-001_T2star_wmseg_erode.nii.gz --mask-wm sub-001_T2star_wmseg.nii.gz --mask-gm sub-001_T2star_gmseg.nii.gz --subject sub-001 --output ./results.csv
```
Note - I have only a single image; thus, I pass an empty variable to `--data2` to be compatible with the syntax; see [here](https://github.com/sct-pipeline/gm-challenge/blob/f8fd2b877ec7abe6f545c98c9a9cdd542c2e2042/process_data.sh#L137)

The `compute_cnr` script fails on the error due to negative values presented in the `sub-001_T2star_wmseg.nii.gz`:

```
/usr/local/lib/gm-challenge/gmchallenge/compute_cnr.py:94: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(variance)
```

Debug:

<img width="700" alt="image" src="https://user-images.githubusercontent.com/39456460/178916468-dc1d8089-1630-453b-92a2-373b91b8d562.png">


## My solution

Threshold the WM segmentation before the erosion:

```
sct_maths -i sub-001_wmseg.nii.gz -thr 0 -o sub-001_wmseg.nii.gz
```

## Possible cause of the bug

After the load of the `sub-001_T2star_wmseg.nii.gz` (which is `uint8` datatype)  by `compute_cnr` script with `nibabel` (line [here](https://github.com/sct-pipeline/gm-challenge/blob/f8fd2b877ec7abe6f545c98c9a9cdd542c2e2042/gmchallenge/compute_cnr.py#L106)), image datatype is changed to `float64`. This step probably introduces these small negative values?

<img width="500" alt="image" src="https://user-images.githubusercontent.com/39456460/178481170-9e1df45b-ab3b-45eb-8957-3b820bfd8523.png">

**Comment by mguaypaq** [comment]
Some initial observations, based on the files provided (thanks @valosekj for the comprehensive steps to reproduce!):
* The two segmentation files `sub-001_T2star_seg.nii.gz` and `sub-001_T2star_gmseg.nii.gz` both correctly contain only 0 and 1 values.
* However, there is a single voxel (at coordinates (216, 216, 7)) which is present in the gray matter segmentation (so, with value = 1), but **not** in the spinal cord segmentation (so, with value = 0). This is the original reason why there's a -1 in the white matter segmentation: it's obtained by subtraction using `sct_maths`, which does what it's told. Further thoughts on this point:
  * If we expect that the gray matter is always within the spinal cord, then that's a bug in between `sct_deepseg_sc` and `sct_deepseg_gm`. But given that we have more than one method of segmenting the spinal cord (for example, `sct_propseg`), I'm not sure that we can guarantee this kind of relationship between segmentations.
  * It might be nice to add the operations `-min` and `-max` to `sct_maths`, which would allow taking the intersection and the union of two 0-1 images.
  * But as you suggest, right now it's possible to use `sct_maths -thr 0` to at least make the white matter mask be a 0-1 image (getting rid of the voxel which is in gmseg outside of seg).
* However, there's definitely something strange happening with `sct_maths -erode`! The eroded white matter segmentation file reports its dtype as uint8, but the actual values seem to be {-0.003921509720385075, 1.000000118277967, -1.0} instead of {0, 1, -1}.

**Comment by mguaypaq** [comment]
Some more observations:
* The provided `.nii.gz` files have almost the same header. The main difference is in the `scl_slope` and `scl_inter` header values (which nibabel doesn't show accurately, sadly).
* The `wmseg.nii.gz` correctly encodes the values 0, 1, -1:
  * it uses `scl_slope = 1.0` and `scl_inter = -1.0`
  * -1 is encoded as `0*scl_slope + scl_inter == -1.0`
  * 0 is encoded as `1*scl_slope + scl_inter == 0.0`
  * 1 is encoded as `2*scl_slope + scl_inter == 1.0`
* The `wmseg_erode.nii.gz` file does **not** encode values correctly:
  * it uses `scl_slope = 0.007843137718737125` and `scl_inter = -1.0`
  * -1 is encoded as `0*scl_slope + scl_inter == -1.0`
  * 0 is (badly) encoded as `127*scl_slope + scl_inter == -0.003921509720385075`
  * 1 is (badly) encoded as `255*scl_slope + scl_inter == 1.000000118277967`

I'll keep digging, but this really looks like a bug with either nibabel or one of its dependencies when encoding data to a file.

**Comment by joshuacwnewton** [comment]
Ah! The mention of `nibabel`'s handling of slope/intercept reminded me of https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3232. Might it be related?

**Comment by joshuacwnewton** [comment]
> Ah! The mention of `nibabel`'s handling of slope/intercept reminded me of #3232. Might it be related?

Following the information from https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3232#issuecomment-782270904, I added debugging lines to `Image.save`,  then ran the commands from https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3843#issuecomment-1184069252:

<details>

Debugging lines:

```python
        dtype_header = self.header.get_data_dtype()
        dtype_data = self.data.dtype
        if dtype_data != dtype_header:
            logger.warning(f"WARNING: Header dtype '{dtype_header}' does not match data dtype '{dtype_data}'.")
```

Console logs:

```console
(venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption$ sct_deepseg_sc -i sub-001_T2star.nii.gz -c t2s

--
Spinal Cord Toolbox (git-jn/3843-fix-corrupted-values-from-registration-7b9e09112a1d907ddbeb71381f684436498ebea1*)

sct_deepseg_sc -i sub-001_T2star.nii.gz -c t2s
--

Config deepseg_sc:
  Centerline algorithm: svm
  Brain in image: False
  Kernel dimension: 2d
  Contrast: t2s
  Threshold: 0.89
Creating temporary folder (/tmp/sct-20221021110924.049538-ql26bz3g)
Reorient the image to RPI, if necessary...
Finding the spinal cord centerline...
Creating temporary folder (/tmp/sct-20221021110924.389119-j_tqw87c)
Remove temporary files...
rm -rf /tmp/sct-20221021110924.389119-j_tqw87c
Cropping the image around the spinal cord...
Normalizing the intensity...
Segmenting the spinal cord using deep learning on 2D patches...
Reassembling the image...
Resampling the segmentation to the native image resolution using linear interpolation...
Binarizing the resampled segmentation...
Compute shape analysis: 100%|################| 10/10 [00:00<00:00, 179.97iter/s]
Remove temporary files...
rm -rf /tmp/sct-20221021110924.049538-ql26bz3g
File /home/joshua/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption/sub-001_T2star_seg.nii.gz already exists. Will overwrite it.

Done! To view results, run one of the following commands (depending on your preferred viewer):
fsleyes /home/joshua/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption/sub-001_T2star.nii.gz -cm greyscale /home/joshua/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption/sub-001_T2star_seg.nii.gz -cm red -a 70.0 &

fslview_deprecated /home/joshua/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption/sub-001_T2star.nii.gz -l Greyscale /home/joshua/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption/sub-001_T2star_seg.nii.gz -l Red -t 0.7 &

fslview /home/joshua/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption/sub-001_T2star.nii.gz -l Greyscale /home/joshua/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption/sub-001_T2star_seg.nii.gz -l Red -t 0.7 &

(venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption$ sct_deepseg_gm -i sub-001_T2star.nii.gz

--
Spinal Cord Toolbox (git-jn/3843-fix-corrupted-values-from-registration-7b9e09112a1d907ddbeb71381f684436498ebea1*)

sct_deepseg_gm -i sub-001_T2star.nii.gz
--


Done! To view results, run one of the following commands (depending on your preferred viewer):
fsleyes sub-001_T2star.nii.gz -cm greyscale -a 100.0 sub-001_T2star_gmseg.nii.gz -cm red -a 70.0 &

fslview_deprecated sub-001_T2star.nii.gz -l Greyscale -t 1 sub-001_T2star_gmseg.nii.gz -l Red -t 0.7 &

fslview sub-001_T2star.nii.gz -l Greyscale -t 1 sub-001_T2star_gmseg.nii.gz -l Red -t 0.7 &

(venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption$ sct_math
s -i sub-001_T2star_seg.nii.gz -sub sub-001_T2star_gmseg.nii.gz -o sub-001_T2star_wmseg.nii.gz
 
--
Spinal Cord Toolbox (git-jn/3843-fix-corrupted-values-from-registration-7b9e09112a1d907ddbeb71381f684436498ebea1*)

sct_maths -i sub-001_T2star_seg.nii.gz -sub sub-001_T2star_gmseg.nii.gz -o sub-001_T2star_wmseg.nii.gz
--

File sub-001_T2star_wmseg.nii.gz already exists. Will overwrite it.
WARNING: Header dtype 'uint8' does not match data dtype 'int64'.

Done! To view results, run one of the following commands (depending on your preferred viewer):
fsleyes sub-001_T2star_wmseg.nii.gz &

fslview_deprecated sub-001_T2star_wmseg.nii.gz &

fslview sub-001_T2star_wmseg.nii.gz &

(venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption$ sct_maths -i sub-001_T2star_wmseg.nii.gz -erode 1 -o sub-001_T2star_wmseg_erode.nii.gz

--
Spinal Cord Toolbox (git-jn/3843-fix-corrupted-values-from-registration-7b9e09112a1d907ddbeb71381f684436498ebea1*)

sct_maths -i sub-001_T2star_wmseg.nii.gz -erode 1 -o sub-001_T2star_wmseg_erode.nii.gz
--

File sub-001_T2star_wmseg_erode.nii.gz already exists. Will overwrite it.
WARNING: Header dtype 'uint8' does not match data dtype 'float64'.

Done! To view results, run one of the following commands (depending on your preferred viewer):
fsleyes sub-001_T2star_wmseg_erode.nii.gz &

fslview_deprecated sub-001_T2star_wmseg_erode.nii.gz &

fslview sub-001_T2star_wmseg_erode.nii.gz &
```

</details>

During the erosion step, the warning is triggered here:

```
File sub-001_T2star_wmseg_erode.nii.gz already exists. Will overwrite it.
WARNING: Header dtype 'uint8' does not match data dtype 'float64'.
```

Notably, this specific datatype mismatch _(header = `int`, data array = `float`)_ is what triggers [float-to-int scaling from nibabel](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3232#issuecomment-782270904).

----

To debug further, I set a breakpoint right before `nib.save()` and looked at the data array of the image being saved:

![image](https://user-images.githubusercontent.com/16181459/197230930-ed0b9178-ace8-48b6-bb5f-ce7054720c39.png)

In other words, the data array for the eroded image is perfectly fine right up until the actual saving operation. 

So, the issue that @mguaypaq observed:

> However, there's definitely something strange happening with sct_maths -erode! The eroded white matter segmentation file reports its dtype as uint8, but the actual values seem to be {-0.003921509720385075, 1.000000118277967, -1.0} instead of {0, 1, -1}.

Is _definitely_ caused by #3232.

**Comment by joshuacwnewton** [comment]
My next goal was to find out where the datatype mismatch was introduced in the first place. 

I've narrowed it down to here, inside the call to `sct_maths -erode`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/7b9e09112a1d907ddbeb71381f684436498ebea1/spinalcordtoolbox/scripts/sct_maths.py#L314

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/7b9e09112a1d907ddbeb71381f684436498ebea1/spinalcordtoolbox/image.py#L412-L414

* The output of `nib.load` is a `Nifti1Image` object with an `ArrayProxy` object of type `uint8`:

    ![image](https://user-images.githubusercontent.com/16181459/197250735-6d17f1d5-cd5b-4465-a024-abc6f7c834ef.png)

* Yet, when `im_file.get_data()` is called, the return data array has type `float64` instead:

    ![image](https://user-images.githubusercontent.com/16181459/197250876-23ab59ed-51be-4330-8614-19224ee7800b.png)

To explain why this happens:

* Internally, `im_file.get_data()` calls `np.asanyarray(self._dataobj)`.
* `np.asanyarray` will use the built-in `__array__` property of objects passed to it. For `ArrayProxy` objects, the [`__array__` property](https://github.com/nipy/nibabel/blob/69307d3ac8f6127a2d1ca0e7f6d480bbf4657af0/nibabel/arrayproxy.py#L370-L394) will call the `self._get_scaled()` method, which then [_uses the `dtype` from `scl_slope`_](https://github.com/nipy/nibabel/blob/69307d3ac8f6127a2d1ca0e7f6d480bbf4657af0/nibabel/arrayproxy.py#L349-L351) if a datatype is not explicitly specified.
* Because `scl_slope` is `1.0`, i.e. float, then the data array gets cast to `float64`.

----

Because `im_in` was loaded incorrectly with a `float64` datatype, the eroded image will also have a `float64` datatype, which then causes the datatype mismatch and value corruption when the eroded image is later saved.

**Comment by joshuacwnewton** [comment]
tl;dr:

* The subtraction causes a value of `-1` in the WM image.
* The presence of a `-1` value causes the `scl_scope` property to be float-valued.
* When the WM image is reloaded, because `scl_slope` is a float, `nibabel` converts the data array from `uint8` to `float64`, even though the header _is still `uint8`_.
* This header-array datatype mismatch then propagates to the eroded image, and triggers #3232 upon saving, causing the corrupted values.

**Comment by joshuacwnewton** [comment]
Considering that https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3854 suggests replacing `get_data()` with `get_fdata()`, I thought that perhaps `get_fdata()` might have more intuitive behavior.

But, `get_fdata` _only_ returns float-valued data arrays, which leaves me even more confused. Wouldn't that just... cause _more_ float-int scaling issues? How does `nibabel` expect us to work with integer-valued NIfTI images if it's trying to coerce data arrays into `float64` wherever possible?

**Comment by joshuacwnewton** [comment]
We could _maybe_ do something like this in [`Image.loadFromPath`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/7b9e09112a1d907ddbeb71381f684436498ebea1/spinalcordtoolbox/image.py#L403-L419):

```python
        im_file = nib.load(self.absolutepath, mmap=mmap)
        self.hdr = im_file.header.copy()
        self.data = im_file.get_fdata().astype(self.hdr.get_data_dtype())
```

That way, we guarantee that data arrays are loaded using the type specified by the image's header, which avoids any sort of datatype mismatch shenanigans. 

<details>
<summary> Demonstration using erode command </summary>

```console
(venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/data/issue_3843_pam50-corruption$ sct_maths -i sub-001_T2star_wmseg.nii.gz -erode 1 -o sub-001_T2star_wmseg_erode.nii.gz

--
Spinal Cord Toolbox (git-jn/3843-fix-corrupted-values-from-registration-7b9e09112a1d907ddbeb71381f684436498ebea1*)

sct_maths -i sub-001_T2star_wmseg.nii.gz -erode 1 -o sub-001_T2star_wmseg_erode.nii.gz
--

File sub-001_T2star_wmseg_erode.nii.gz already exists. Will overwrite it.

Done! To view results, run one of the following commands (depending on your preferred viewer):
fsleyes sub-001_T2star_wmseg_erode.nii.gz &

fslview_deprecated sub-001_T2star_wmseg_erode.nii.gz &

fslview sub-001_T2star_wmseg_erode.nii.gz &
```

No datatype mismatch warning! The issues from https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3843#issuecomment-1287244249 are completely mitigated.

</details>

Yet, this also has the downside of coercing the `int64` `(0, 1, -1)` array into a `uint8` `(0, 1)` array which is... probably not good? While it does match expected results, it also hides the fact that something fishy happened with the `-sub` step. :thinking:

**Comment by joshuacwnewton** [comment]
Given that this issue is a combination of:

* Intended usage (i.e. the `-1` was introduced by a properly-functioning `-sub` operation).
* Issue [#3232](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3232).

I'd like to propose closing this issue as resolved, and continuing discussion in #3232.
